test/functional/apm_logging_tests.js

```diff
@@ -0,0 +1,249 @@
+        done(err);
+      };
+      runTestClient(this, '?monitor=command&monitorOut=apm_test_log', cb);
+    });
+
+    // skip for now: this test depends on the new fixes in the core uri parser
+    it.skip('log both command and sdam events using `monitor=command,sdam`', function(done) {
+      const cb = (err, data) => {
+        expect(data.numLines).to.equal(8);
+        expect(data.sdam).to.not.be.empty;
+        expect(data.command).to.not.be.empty;
+        done(err);
+      };
+      runTestClient(this, '?monitor=command,sdam&monitorOut=apm_test_log', cb);
+    });
+
+    it('log both command and sdam events using `monitor=all`', function(done) {
+      const cb = (err, data) => {
+        expect(data.numLines).to.equal(8);
+        expect(data.sdam).to.not.be.empty;
+        expect(data.command).to.not.be.empty;
+        done(err);
+      };
+      runTestClient(this, '?monitor=all&monitorOut=apm_test_log', cb);
+    });
+
+    it('should append timestamp to a pre-existing log filename to avoid overwriting', function(done) {
+      // suppress and spy on the renaming function
+      sinon.stub(fs, 'renameSync');
+      // trick the code into thinking log files already exist
+      sinon.stub(fs, 'existsSync').callsFake(() => {
+        return true;
+      });
+      const testStartStamp = new Date();
+      const cb = err => {
+        const renameArgs = fs.renameSync.lastCall.args;
+        const fileTimestamp = new Date(renameArgs[1].substring(renameArgs[1].length - 24));
+        expect(renameArgs[0]).to.equal(testLogFileName);
+        // ensure the date was appended to the file at the same time or after this test started
+        expect(fileTimestamp).to.be.at.least(testStartStamp);
+        fs.existsSync.restore();
+        fs.renameSync.restore();
+        done(err);
+      };
+      runTestClient(this, '?monitor=command&monitorOut=apm_test_log', cb);
+    });
+  });
+
+  describe('Error handling', function() {
+    let writeStream;
+
+    before(function() {
+      writeStream = new stream.Writable();
+      sinon.stub(writeStream, 'write');
+
+      // give us access to the write stream in order to manipulate it
+      sinon.stub(fs, 'createWriteStream').callsFake(() => {
+        return writeStream;
+      });
+
+      return setupDatabase(this.configuration);
+    });
+
+    after(function() {
+      fs.createWriteStream.restore();
+    });
+
+    it('should call callback with error when log filename cannot be resolved', function(done) {
+      const configuration = this.configuration;
+      const url = configuration.url() + '?monitor=all&monitorOut=invalid:name';
+      const client = new MongoClient(url, { w: 1, useNewUrlParser: true });
+
+      client.connect(function(err) {
+        expect(err).to.exist;
+        expect(err).to.be.an.instanceOf(TypeError);
+        done();
+      });
+    });
+
+    it('should throw error and close client when file stream emits error after client has connected', function(done) {
```

Comment:
This test relies on `client.close` already being stubbed from the `session_leak_test.js` [`beforeEach` hook](https://github.com/mongodb/node-mongodb-native/blob/78f69774ea45fb1e130eebb7e4bce54205296a8f/test/functional/session_leak_test.js#L59). Without that hook being run (like if just this test file were to be run by itself), this test fails, since in that case `client.close` will not already be stubbed. Is this acceptable? The test should always pass when the entire suite is being run.

Reply:
Can this be solved with something like `stub.callThrough();`? Alternatively, if sinon is hindering us here we can always stub it manually and call the original function.